overlord/configstate/config/helpers.go

```diff
@@ -101,15 +101,24 @@ func PatchConfig(snapName string, subkeys []string, pos int, config interface{},
 		if err := jsonutil.DecodeWithNumber(bytes.NewReader(*config), &configm); err != nil {
 			return nil, fmt.Errorf("snap %q option %q is not a map", snapName, strings.Join(subkeys[:pos], "."))
 		}
-		_, err := PatchConfig(snapName, subkeys, pos, configm, value)
+		result, err := PatchConfig(snapName, subkeys, pos, configm, value)
 		if err != nil {
 			return nil, err
 		}
-		return jsonRaw(configm), nil
+
+		// PatchConfig may have recreated higher level element that was previously set to null in same
+		// transaction; returning the result for PatchConfig rather than just configm ensures we do
+		// support cases where a previously unset path is set back.
+		return jsonRaw(result), nil
 
 	case map[string]interface{}:
 		// Update map to apply against pristine on commit.
 		if pos+1 == len(subkeys) {
+			// we may encounter nil when setting back values that were previously set to null in
+			// same transaction (see LP: #1920773).
+			if config == nil {
```

Comment:
this is ok but it means we sometimes store   (interface{})(nil) and sometimes (map[string]interface{})(nil), can we know how that happens? (maybe s very natural, unavoidable but wondering),  if we always used the former we would hit the first case of the switch

Reply:
It's the result of going through `case *json.RawMessage:` in PatchConfig when value is null (json.RawMessage encoding of null) - we decode it temporarily to `configm map[string]interface{}`.

Reply:
AFAICT the bare nil happens when we patch the config where the current key does not exist in the map[string]interface{}, eg.
```
m := map[string]interface{}{}
jv := json.RawMessage("1")
PatchConfig("foo", []string{"foo","bar","baz", m, 0, &jv)
```
will hit:
- case map[string]interface{} for pos 0
- case interface{} for pos 1 as m["foo"] == nil // interface{}
- case map[string]interface{} for pos 1 as we created an empty map
- case interface{} for pos 2 
and so on.

I think we can drop the nil case if we do something like this:
```
...
	case map[string]interface{}:
		// Update map to apply against pristine on commit.
		...
		if pos+1 == len(subkeys) {
			...
		} else {
			if config[subkeys[pos]] == nil {
					config[subkeys[pos]] = make(map[string]interface{})
			}
			result, err := PatchConfig(snapName, subkeys, pos+1, config[subkeys[pos]], value)
			if err != nil {
				return nil, err
			}
			config[subkeys[pos]] = result
			return config, nil
		}
	}
```

Reply:
I don't think this code is much clearer that having the `case nil

Reply:
yea, the issue then though is that PatchConfig takes a interface{}  and expects to handle the null/nil situation in the first case. We are kind of repeating that handling here.  My preference would be to be more explicit about this,  also keep to the invariant that we tend to pass and return type matching arguments if possible. So I tried something different and it seems to work as well:

https://gist.github.com/pedronis/a12e5f4c3f5e14b283b079d4270a5f9e

Reply:
> AFAICT the bare nil happens when we patch the config where the current key does not exist in the map[string]interface{}

no, it can happen also when we put an explicit nil or *json.RawMessage("null") somewhere

Reply:
> yea, the issue then though is that PatchConfig takes a interface{} and expects to handle the null/nil situation in the first case. We are kind of repeating that handling here. My preference would be to be more explicit about this, also keep to the invariant that we tend to pass and return type matching arguments if possible. So I tried something different and it seems to work as well:
> 
> https://gist.github.com/pedronis/a12e5f4c3f5e14b283b079d4270a5f9e

I like that, applied, thank you!